realtek: pcs: rtl931x: SerDes initialization refactoring and fixes

jonasjelonek opened the pull request:
Fix some issues in RTL931X SerDes initialization. Most importantly, this adopts the patch done by @0xharshal and @ecsv (provided in [1]) to make it work in the PCS driver.

Tested on Zyxel XS1930-10 with following observations:
- 10GBase-LR, 10GBase-SR working fine
- 1000Base-SX working fine
- 1000Base-T (SGMII) fine on 1 of 2 devices
- 10G DAC: still issues with link flapping but works in theory

I left out the USXGMII fixes from the patch because USXGMII (+ XSGMII) initialization needs a lot more love to be properly implemented.

[1] https://github.com/openwrt/openwrt/pull/19579#issuecomment-3371818370

jonasjelonek commented:
@0xharshal Did you do some research on your patch regarding which of the SerDes magic writes from the SDK we actually need and which not? I've seen a lot of extra stuff e.g. [here](https://github.com/plappermaul/realtek-doc/blob/69d2890a2e2d7a03df6e40e2cd2c32ff5b074dcf/sources/rtk-dms1250/src/hal/phy/phy_rtl9310.c#L2302). Using this extra stuff works too but I'm not sure if we should leave it out or not.

0xharshal commented:
Hey @jonasjelonek I got sick today and taking some rest until monday. Will be back on tuesday and then try out these changes for sure.

jonasjelonek commented:
Let's see how well that works for now but it's probably not the final state. While this works fine on my Zyxel XS1930-10, SGMII does not (yet) work properly on my XikeStor SKS8300-12X.

DAC cables are still causing issues sometimes with this. It basically works but sometimes the cables have some kind of high frequent link flapping.

0xharshal commented:
> Ok, that is strange. Unless I missed something, with the diff this should be more or less what was working for Sven/you. Not sure if something has extremely changed since then.
> 
Well I was not following newer developments apart for backporting serdes register debugging to kernel 5.15 and RTLSDK and  currently busy with making link aggregation upstreamable. I need some time to figure out what is the breaking change.

> Is the link flapping really extreme so you don't get a consistent link or does it work sometimes? Any other behaviors which might be important?
> 

Yes, for me link goes down so quickly that I am not yet able to see "UP" in ip link. only dmesg output shows that link is up and then immediately goes down. And I also see corresponding LED to light up and quickly turn off. 

> I experienced link flapping on my XikeStor with DAC cables, and it got worse/better with removing/applying some register writes from the SDK which I think might me calibration values.

Hmm... If you have access to u-boot on your switch can you try following command and see if the DAC cable works consistently in u-boot.

```
rtk 10g [PORT] dac50cm
# or
rtk 10g [PORT] dac100cm
# or
rtk 10g [PORT] dac300cm
```
If it works in the u-boot, you can atleast figure out the missing serdes calibration early during the linux boot.

0xharshal commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -2320,7 +2364,81 @@ static sds_config sds_config_10p3125g_cmu_type1[] = {
+	/* from _phy_rtl9310_10gMedia_set */
+
+	/* gating of ? */
+	rtpcs_sds_write_bits(ctrl, sds, 0x5f, 0x1, 0, 0, 0x1);
```
Yep, you are right. I am trying to implement calibration routines on top of this PR and see if situation improves. couldn't make sense of some stuff so I asked here.

0xharshal commented:
This contains my changes,  currently not organized at all. https://github.com/0xharshal/openwrt/commits/realtek/pcs-serdes/

So far current issues on PSX28 with my branch,
#### All SFP ports work but behave like they are hanging by a thread.
10g fiber is plugged in to port 28 (serdes 11, internal port 53), on boot it is turned on and works as expected. Without touching that, If i plug in a 1G Copper SFP in to port 27 (serdes 10, internal port 52) the 10g fiber looses connectivity and does not recover until i remove and plug in fiber cable again.

#### Switch between 1g/10g mode does not work all the time
#### high cpu usage reported due to using single thread work-queue with mdelay()
#### No shared locking between link_check work-queue and serdes-configuration routines

Can you test this on your device and see if anything improves/worsens?

0xharshal commented:
> Probably it'll be easier then to fix such outstanding issues in smaller steps.

Agreed. It does not break any thing which is already not broken.

jonasjelonek committed:
realtek: pcs: rtl931x: fix clearing symbol errors

Fix the implementation to clear symbol errors. Up to now, it was barely
functional because important modes weren't handled. Add another function
(which is needed in the future anyway) and pick to missing bits from
that function into the symbol clearing function to handle the modes
1000Base-X and 10GBaseR too. All that is based on [1].

While at it, rename the function to adhere to a common naming scheme and
fix some minor style issues.

[1] https://github.com/plappermaul/realtek-doc/blob/69d2890a2e2d7a03df6e40e2cd2c32ff5b074dcf/sources/rtk-dms1250/src/hal/phy/phy_rtl9310.c#L703

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl931x: reorder functions

Bring the RTL931X functions into a proper order for two purposes:
- be able to reuse code
- have blocks of logically connected functions

This just moves code, no functional changes.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl931x: merge helper into mode setting

Merge the 'rtpcs_931x_sds_disable' helper into the MII mode setting and
use that instead. The helper was essentially doing same just with a
value for OFF state.

Moreover, the name of the helper was confusing. It implied it disables
the whole SerDes. However, this is used in Fiber mode setting and thus,
cannot completely disable the SerDes.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl931x: use hardware mode for mode setting

Use the hardware mode instead of PHY_INTERFACE_MODE_* mode for mode
setting. Adjust all affected functions and switch cases accordingly.

In MII mode setting, drop the case for 2500Base-X as this is a mistake.
In the SDK reference code, this doesn't exist [1].

[1] https://github.com/plappermaul/realtek-doc/blob/69d2890a2e2d7a03df6e40e2cd2c32ff5b074dcf/sources/rtk-dms1250/src/hal/phy/phy_rtl9310.c#L817

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl931x: merge and drop unused helper

Merge the unused helper 'rtpcs_931x_sds_fiber_disable' into Fiber mode
setting, and drop the helper itself. As with the MII helper in a
previous commit, functionality is basically the same just with a value
for OFF mode. If functionality is required later, Fiber mode setting can
be used with the OFF mode instead of carrying the unused helper until
it's used.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl931x: fix MII mode setting

The function 'rtpcs_931x_sds_mii_mode_set' does not correctly write the
register. It just write a plain value at the determined register
address. While this works for SerDes with (id mod 4 == 0), it doesn't
for the other SerDes.

Fix that by using a corresponding shift and writing only some bits
instead of the whole register.

While at it, drop an unneeded blank line, add comment to explain a bit
that is set and use the BIT(..) helper for that instead of manual shift.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl931x: improve mode setting functions

Improve the current mode setting functions to address a few issues:
- add missing mode from SDK reference (instead of using the
  corresponding value as a default value)
- use and return error values
- give internal variable a more meaningful name
- consistently use lowercase letters in hex values

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl931x: streamline hardware mode setting

The SDK and our code for finally applying the hardware mode are quite
confusing. There are two different "places" where a mode can be set,
in a SerDes register and in a global SerDes mode register. Neither the
SDK nor any of the datasheet/documentation serve any explanation for
that. The functions are just named "fiber_mode_set" and "mii_mode_set"
which is basically as useless as it can be to understand it.

Try to get rid off this confusion by naming the functions
'sds_set_mac_mode' and 'sds_set_ip_mode' to make clear where the mode
is set. While at it, also clarify the naming of 'config_mode' by
renaming it to 'config_hw_mode'.

The naming is based on the following assumption:

> Realtek uses an SerDes IP core design (probably from another vendor)
> in their switch. This supports a variety of modes and must be
> configured properly for each mode. Usually, changing the mode in the
> MAC's registers triggers a proper configuration of the SerDes IP block
> in the background.

> However, for some modes this seems to be incomplete, at least missing
> important parts so it doesn't work on its own in the end. In this
> case, the SerDes IP block needs to be configured manually with the
> missing bits to make it work.

There are several places in the SDK that support this assumption, both
for RTL931X and RTL930X (as they are somewhat similar), e.g. [1].

[1] https://github.com/plappermaul/realtek-doc/blob/f7f85ffc14c6e575ad94f2b427106de533b888fe/sources/rtk-dms1250/src/dal/longan/dal_longan_sds.c#L1746

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl931x: fix SerDes mode application

The SerDes mode setting at the end of rtpcs_931x_setup_serdes is
currently broken although it is mostly similar to what the SDK does [1].
It prevents several modes from being set, especially fiber modes. This
seems to be one (if not the only) issue for currently missing SFP
support.

Add a small helper function which takes cares of setting the mode, to
keep the still valid different procedure when using XSGMII mode. Only
this helper is called in rtpcs_931x_setup_serdes to keep it clean there.

As a functional change, call mode application in every case, not just
for SGMII, QSGMII and USXGMII. We can assume the SDK is misleading in
this case, either accidentially or on purpose. This makes SFP modules
work in theory. In practice, there still seem to be device-specific
issues which need to be fixed later. These issues may include no link
detection or link flapping.

[1] https://github.com/plappermaul/realtek-doc/blob/f7f85ffc14c6e575ad94f2b427106de533b888fe/sources/rtk-dms1250/src/dal/mango/dal_mango_construct.c#L2266

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/20736
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek commented:
@plappermaul Please review with a critical eye as usual. And if possible, test with your LGS352C. This is only the beginning of making SFP on RTL931x work so I don't expect it to work on your device. During the discussion here, there still seems to be a lot missing to make it work consistently on all devices. But at least it works on some of my devices.

I'd be interested on your opinion about c452571, wearing my tin foil hat I made an assumption about why there are two places to set the SerDes mode. If that's totally off, I can just remove that.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -2340,30 +2340,61 @@ static void rtpcs_931x_sds_disable(struct rtpcs_serdes *sds)
 		break;
+	case PHY_INTERFACE_MODE_1000BASEX:
+	case PHY_INTERFACE_MODE_10GBASER:
+		rtpcs_931x_sds_fiber_get_symerr(sds, mode);
```
I always disliked the fact that the SDK uses different levels of indirection & subfunctions for different modes. So either call subfunctions each times or program registers each time. Additionaly let the clear/get functions do only what they are named after. If really needed call get_symerr & clear_symerr from the initial location subsequently.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -2387,6 +2387,9 @@ static void rtpcs_931x_sds_mii_mode_set(struct rtpcs_serdes *sds,
 		val = 0x6;
 		break;
+	case RTPCS_SDS_MODE_HISGMII:
+		val = 0x12;
```
While you are here give this variable a meaningful name.

jonasjelonek replied:
done

plappermaul reviewed (changes requested):
The RLT8218D ports on my LGS352C still work.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -2476,6 +2476,15 @@ static int rtpcs_931x_sds_set_ip_mode(struct rtpcs_serdes *sds,
+		return rtpcs_931x_sds_set_mac_mode(sds, hw_mode);
+	else
+		return rtpcs_931x_sds_set_ip_mode(sds, hw_mode);
+}
```
This comment is not a change request, only a friendly reminder. At this moment we have

- a mac_set_mode that covers several modes (e.g. A, B, F, G)
- a ip_set_mode that covers several modes (e.g. B, D, F, H)
- And this wrapper that decides which setter to call

At some point in time we should get a clear view, where/how we get these three conditions best aligned. There are two options possible.

- mac_mode covers all modes and can be called from ip_mode if required
- mac_mode covers only modes that can be fully automatically handled 

Future will tell ...

jonasjelonek replied:
noted and acked, thanks!

plappermaul reviewed (changes requested):
Thanks for this work. Having working SFP+ slots on XGS1930 is a great result.

plappermaul reviewed (approved):
Medal of "pioneer" earned. Thank you.

jonasjelonek commented:
@robimarko finally, this should be ready to go. Thanks in advance!

robimarko commented:
Thanks! Rebased on top of main and merged!
